### PR TITLE
[DROOLS-6556] JSON marshalling slow due to inefficient com.fasterxml.…

### PR DESCRIPTION
**Pull request by tkobayas**
…jackson.databind.type.TypeFactory.findClass calls

- WIP

**JIRA**: 
https://issues.redhat.com/browse/DROOLS-6556

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
 
* <b>a full downstream build</b> please add comment: <b>Jenkins run fdb</b>
  
* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java [hunk 1]
```diff
@@ -105,6 +105,10 @@
 
     private boolean fallbackClassLoaderEnabled = Boolean.parseBoolean(System.getProperty("org.kie.server.json.fallbackClassLoader.enabled", "false"));
 
+    private boolean findDeserializerFirst = Boolean.parseBoolean(System.getProperty("org.kie.server.json.findDeserializerFirst.enabled", "true"));
```
Comment (tkobayas) [inline kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java:108]: Set `true` for `findDeserializerFirst` to see if all test cases pass.

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java [hunk 2]
```diff
@@ -938,6 +945,16 @@ public Object deserializeTypedFromObject(JsonParser jp, DeserializationContext c
...
+                    if (findDeserializerFirst) {
+                        String nextFieldName = jp.nextFieldName();
+//                        logger.info("nextFieldName : {}", nextFieldName);
+                        if (!baseTypeName().equals(nextFieldName)) {
+                            JsonDeserializer<Object> deser = _findDeserializer(ctxt, baseTypeName());
+                            Object value = deser.deserialize(jp, ctxt);
+                            return value;
+                        }
+                    }
+
```
Comment (tkobayas) [inline kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java:957]: `jp.nextFieldName()` proceeds the current token but `_findDeserializer()` and `super.deserializeTypedFromObject` can handle it.

**Comment by tkobayas** [comment]
Benchmark results with https://github.com/tkobayas/kie-benchmarks/blob/DROOLS-6556/simple-benchmarks/src/main/java/org/drools/benchmarks/marshaller/JSONMarshallerDroolsCommandBenchmark.java
```
Benchmark                                  (allClassNamePayload)  (findDeserializerFirst)  Mode  Cnt    Score   Error  Units
JSONMarshallerDroolsCommandBenchmark.test                  false                    false    ss  100  309.270 ± 2.267  ms/op
JSONMarshallerDroolsCommandBenchmark.test                  false                     true    ss  100   11.168 ± 0.954  ms/op
JSONMarshallerDroolsCommandBenchmark.test                   true                    false    ss  100   14.598 ± 1.087  ms/op
JSONMarshallerDroolsCommandBenchmark.test                   true                     true    ss  100   14.662 ± 1.179  ms/op
```
This means:
- If all class names are not written in the  payload (= current kie-server-client Drools Command), lots of Class.forName() are called (= can block threads) and lots of ClassNotFoundException happen (= time consuming). So `findDeserializerFirst` significantly improve the performance.
- If all class names are written in the  payload, Class.forName() is not called a lot (deserializer cache seems to work). So performance is good. The difference from `allClassNamePayload == false && findDeserializerFirst == true` may be explained by the size of the payload.

**Comment by tkobayas** [comment]
Jenkins retest this

**Comment by tkobayas** [comment]
Jenkins retest this

**Comment by mareknovotny** [comment]
so kie-server-integ-tests-optaplanner IT failed due not able to start container as there were some leaked not stopped WF container?
`[2021-08-30T03:43:02.052Z] [INFO] [talledLocalContainer] 23:43:02,179 ERROR [org.jboss.msc.service.fail] (MSC service thread 1-5) MSC000001: Failed to start service org.wildfly.management.http.extensible: org.jboss.msc.service.StartException in service org.wildfly.management.http.extensible: java.net.BindException: Address already in use /127.0.0.1:43646`

**Comment by tkobayas** [comment]
@mareknovotny 
It doesn't look to be "leaked not stopped WF"
```
$ grep -E 'WFLYSRV0025|WFLYSRV0050' consoleText 
[2021-08-30T02:26:22.055Z] [INFO] [talledLocalContainer] 22:26:21,928 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0025: WildFly Full 23.0.2.Final (WildFly Core 15.0.1.Final) started in 4961ms - Started 60 of 86 services (35 services are lazy, passive or on-demand)
[2021-08-30T02:26:23.139Z] [INFO] [talledLocalContainer] 22:26:23,087 INFO  [org.jboss.as] (MSC service thread 1-1) WFLYSRV0050: WildFly Full 23.0.2.Final (WildFly Core 15.0.1.Final) stopped in 44ms
[2021-08-30T02:27:16.981Z] [INFO] [talledLocalContainer] 22:27:16,989 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0025: WildFly Full 23.0.2.Final (WildFly Core 15.0.1.Final) started in 45382ms - Started 592 of 792 services (375 services are lazy, passive or on-demand)
[2021-08-30T02:29:27.246Z] [INFO] [talledLocalContainer] 22:29:27,204 INFO  [org.jboss.as] (MSC service thread 1-4) WFLYSRV0050: WildFly Full 23.0.2.Final (WildFly Core 15.0.1.Final) stopped in 2092ms
[2021-08-30T02:29:54.098Z] [INFO] [talledLocalContainer] 22:29:54,036 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0025: WildFly Full 23.0.2.Final (WildFly Core 15.0.1.Final) started in 4438ms - Started 60 of 86 services (35 services are lazy, passive or on-demand)
[2021-08-30T02:29:55.527Z] [INFO] [talledLocalContainer] 22:29:55,624 INFO  [org.jboss.as] (MSC service thread 1-6) WFLYSRV0050: WildFly Full 23.0.2.Final (WildFly Core 15.0.1.Final) stopped in 119ms
[2021-08-30T02:30:46.773Z] [INFO] [talledLocalContainer] 22:30:46,707 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0025: WildFly Full 23.0.2.Final (WildFly Core 15.0.1.Final) started in 42080ms - Started 592 of 792 services (375 services are lazy, passive or on-demand)
[2021-08-30T02:56:52.763Z] [INFO] [talledLocalContainer] 22:56:52,707 INFO  [org.jboss.as] (MSC service thread 1-4) WFLYSRV0050: WildFly Full 23.0.2.Final (WildFly Core 15.0.1.Final) stopped in 1003ms
[2021-08-30T02:57:42.053Z] [INFO] [talledLocalContainer] 22:57:42,127 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0025: WildFly Full 23.0.2.Final (WildFly Core 15.0.1.Final) started in 3829ms - Started 60 of 86 services (35 services are lazy, passive or on-demand)
[2021-08-30T02:57:43.145Z] [INFO] [talledLocalContainer] 22:57:43,230 INFO  [org.jboss.as] (MSC service thread 1-1) WFLYSRV0050: WildFly Full 23.0.2.Final (WildFly Core 15.0.1.Final) stopped in 125ms
[2021-08-30T02:58:35.126Z] [INFO] [talledLocalContainer] 22:58:35,247 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0025: WildFly Full 23.0.2.Final (WildFly Core 15.0.1.Final) started in 43341ms - Started 668 of 869 services (380 services are lazy, passive or on-demand)
[2021-08-30T03:42:18.914Z] [INFO] [talledLocalContainer] 23:42:19,020 INFO  [org.jboss.as] (MSC service thread 1-7) WFLYSRV0050: WildFly Full 23.0.2.Final (WildFly Core 15.0.1.Final) stopped in 1184ms
[2021-08-30T03:42:40.233Z] [INFO] [talledLocalContainer] 23:42:40,166 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0025: WildFly Full 23.0.2.Final (WildFly Core 15.0.1.Final) started in 3975ms - Started 60 of 86 services (35 services are lazy, passive or on-demand)
[2021-08-30T03:42:41.280Z] [INFO] [talledLocalContainer] 23:42:41,248 INFO  [org.jboss.as] (MSC service thread 1-4) WFLYSRV0050: WildFly Full 23.0.2.Final (WildFly Core 15.0.1.Final) stopped in 155ms
[2021-08-30T03:43:25.368Z] [INFO] [talledLocalContainer] 23:43:25,473 INFO  [org.jboss.as] (MSC service thread 1-6) WFLYSRV0050: WildFly Full 23.0.2.Final (WildFly Core 15.0.1.Final) stopped in 64ms
```

BindException was raised even though `43646` was reserved by build-helper-maven-plugin:`reserve-network-port`. Do you have any ideas on this?
```
[2021-08-30T03:42:26.087Z] [INFO] --- build-helper-maven-plugin:3.0.0:reserve-network-port (reserve-network-port) @ kie-server-integ-tests-optaplanner ---
[2021-08-30T03:42:26.087Z] [INFO] Reserved port 34163 for container.port
[2021-08-30T03:42:26.087Z] [INFO] Reserved port 40459 for cargo.jboss.ajp.port
[2021-08-30T03:42:26.087Z] [INFO] Reserved port 39076 for cargo.jboss.https.port
[2021-08-30T03:42:26.087Z] [INFO] Reserved port 43646 for cargo.jboss.management-http.port
[2021-08-30T03:42:26.087Z] [INFO] Reserved port 42131 for cargo.jboss.management-https.port

...

[2021-08-30T03:43:02.052Z] [INFO] [talledLocalContainer] 23:43:02,179 ERROR [org.jboss.msc.service.fail] (MSC service thread 1-5) MSC000001: Failed to start service org.wildfly.management.http.extensible: org.jboss.msc.service.StartException in service org.wildfly.management.http.extensible: java.net.BindException: Address already in use /127.0.0.1:43646
[2021-08-30T03:43:02.052Z] [INFO] [talledLocalContainer]        at org.jboss.as.server.mgmt.UndertowHttpManagementService.start(UndertowHttpManagementService.java:399)
[2021-08-30T03:43:02.052Z] [INFO] [talledLocalContainer]        at org.jboss.msc.service.ServiceControllerImpl$StartTask.startService(ServiceControllerImpl.java:1739)
[2021-08-30T03:43:02.052Z] [INFO] [talledLocalContainer]        at org.jboss.msc.service.ServiceControllerImpl$StartTask.execute(ServiceControllerImpl.java:1701)
[2021-08-30T03:43:02.052Z] [INFO] [talledLocalContainer]        at org.jboss.msc.service.ServiceControllerImpl$ControllerTask.run(ServiceControllerImpl.java:1559)
[2021-08-30T03:43:02.052Z] [INFO] [talledLocalContainer]        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
[2021-08-30T03:43:02.052Z] [INFO] [talledLocalContainer]        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1990)
[2021-08-30T03:43:02.052Z] [INFO] [talledLocalContainer]        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
[2021-08-30T03:43:02.052Z] [INFO] [talledLocalContainer]        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1363)
[2021-08-30T03:43:02.052Z] [INFO] [talledLocalContainer]        at java.lang.Thread.run(Thread.java:748)
```

**Comment by mariofusco** [comment]
@mareknovotny @tkobayas I don't see how that test failure could be related with this pull request anyway? Am I missing something? If not, could we merge it anyway? My understanding is that this fix is quite urgent.

**Comment by mareknovotny** [comment]
this is not related, that is true, but due this error the followup modules and tests are not ran. @mariofusco

**Comment by tkobayas** [comment]
Build Chain / Maven Build (8) (pull_request) indeed passed all tests.
https://github.com/kiegroup/droolsjbpm-integration/pull/2577/checks?check_run_id=3441232126
So I think it's good to merge. WDYT? @mareknovotny

**Comment by mareknovotny** [comment]
> @mareknovotny
> It doesn't look to be "leaked not stopped WF"
> 

> 
> BindException was raised even though `43646` was reserved by build-helper-maven-plugin:`reserve-network-port`. Do you have any ideas on this?
> 
> ```
> [2021-08-30T03:42:26.087Z] [INFO] --- build-helper-maven-plugin:3.0.0:reserve-network-port (reserve-network-port) @ kie-server-integ-tests-optaplanner ---
> [2021-08-30T03:42:26.087Z] [INFO] Reserved port 34163 for container.port
> [2021-08-30T03:42:26.087Z] [INFO] Reserved port 40459 for cargo.jboss.ajp.port
> [2021-08-30T03:42:26.087Z] [INFO] Reserved port 39076 for cargo.jboss.https.port
> [2021-08-30T03:42:26.087Z] [INFO] Reserved port 43646 for cargo.jboss.management-http.port
> [2021-08-30T03:42:26.087Z] [INFO] Reserved port 42131 for cargo.jboss.management-https.port
> ...
org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1363)
> [2021-08-30T03:43:02.052Z] [INFO] [talledLocalContainer]        at java.lang.Thread.run(Thread.java:748)
> ```
@sutaakar WDYT ?

**Review by elguardian — APPROVED** [review]

**Comment by sutaakar** [comment]
Looks like a "common" port clash, not sure why. Maybe `reserve-network-port` doesn't work properly.

File: kie-server-parent/kie-server-api/src/test/java/org/kie/server/api/model/DroolsCommandMarshallingTest.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+    @Test
+    public void testJSON() {
+        Marshaller marshaller = MarshallerFactory.getMarshaller(getCustomClasses(), MarshallingFormat.JSON, getClass().getClassLoader());
+        verifyMarshallingRoundTrip(marshaller, createTestCommand(), jsonExpectedPayload);
+    }
+
+    private void verifyMarshallingRoundTrip(Marshaller marshaller, BatchExecutionCommand inputCommand, String expectedPayload) {
+        String rawContent = marshaller.marshall(inputCommand);
+        logger.info(rawContent);
```
Comment (sutaakar) [inline kie-server-parent/kie-server-api/src/test/java/org/kie/server/api/model/DroolsCommandMarshallingTest.java:159]: Can you please change the log level to debug?

File: kie-server-parent/kie-server-api/src/test/java/org/kie/server/api/model/DroolsCommandWithListMarshallingTest.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+    @Test
+    public void testJSON() {
+        Marshaller marshaller = MarshallerFactory.getMarshaller(getCustomClasses(), MarshallingFormat.JSON, getClass().getClassLoader());
+        verifyMarshallingRoundTrip(marshaller, createTestCommand(), jsonExpectedPayload);
+    }
+
+    private void verifyMarshallingRoundTrip(Marshaller marshaller, BatchExecutionCommand inputCommand, String expectedPayload) {
+        String rawContent = marshaller.marshall(inputCommand);
+        logger.info(rawContent);
```
Comment (sutaakar) [inline kie-server-parent/kie-server-api/src/test/java/org/kie/server/api/model/DroolsCommandWithListMarshallingTest.java:220]: Same here.

**Review by sutaakar — APPROVED** [review]
Minor logging related comments, otherwise looks good.

File: kie-server-parent/kie-server-api/src/test/java/org/kie/server/api/model/DroolsCommandMarshallingTest.java (hunk 3 above)
Comment (tkobayas) [reply kie-server-parent/kie-server-api/src/test/java/org/kie/server/api/model/DroolsCommandMarshallingTest.java:159]: @sutaakar Sure, I'll do it in another JIRA.

Comment (tkobayas) [reply kie-server-parent/kie-server-api/src/test/java/org/kie/server/api/model/DroolsCommandMarshallingTest.java:159]: Filed https://issues.redhat.com/browse/DROOLS-6577

**Pull request merged**